cilium/CFP-32820-improve-scalability-of-health-probing.md

```diff
@@ -0,0 +1,149 @@
+
+#### HTTP Probe
+
+The HTTP probe runs in a for loop and [probes each IP](https://github.com/cilium/cilium/blob/fac90aa858c86eefd2c4bdccef8b56abf24a7b58/pkg/health/server/prober.go#L349). Again, for consistency in design, we may modify this probe using [time/rate](https://pkg.go.dev/golang.org/x/time/rate) to run the probe in accordance with the configured requests per second.
+
+### Other Details
+
+#### Connectivity Status Cache Update
+The connectivity cache is currently updated on the condition that the current health report start time is before the health report being cached, at the end of the ProbeInterval period.
+
+With the ProbeInterval set to longer than 60s, it may make more sense to update the connectivity cache more frequently than the ProbeInterval, so that users see some refreshed metrics based on what has been collected at a given point.
+
+A possible solution for this may be to continue refreshing the connectivity cache at intervals of 60 seconds. Instead of storing the start time for the health report as a whole, we may store the health report start time on a per-node basis. Then, every 60 seconds, we make a call to refresh the cache, and update metrics for nodes iff the start time for the node’s last stored metric is before the start time of the current health report.
+
+#### Metrics Update Period
+
+As a consequence of this proposal, health probe updates may take longer and be more unpredictably-timed for users. In order to keep users aware of how long it will take to receive accurate metrics for the known nodes in the cluster, we can update the output of `cilium-health status` to also report the current probe interval and/or the most recent timestamp at which metrics for all known nodes were updated.
+
+#### Probe Timeouts and Errors
+
+The HTTP probe is currently set to have a timeout of 30 seconds. The ICMP probe has MaxRTT set to 60 seconds, which acts as the timeout for the pings. Both timeouts seem larger than necessary. The HTTP probe timeout can be re-configured to 10 seconds, which is a reasonable timeout and accounts for the sizes of the maximum bucket configured for the histogram metric `node_health_connectivity_latency_seconds`.
```

Comment:
Maybe let me turn this feedback upside-down: I think it would be a non-goal for 5,000 nodes with no connectivity to take 50,000s (half a day) to ascertain that no other node is reachable. The probing should complete in around the target probing interval, +/- a few seconds or so.

I _suspect_ that there may be some optimizations to be made around exactly how socket and goroutine management work here, but I think that's also probably too much detail to lock down in the design here. We'll figure out what makes sense in that regard as you develop the implementation.

Reply:
Ah, I was not envisioning taking 10s per node for sure 😅 The second option is more what I was thinking, with ~5.3 pings/second. The probing interval should not be significantly affected by failures -- it should at most be `probingInterval+maxRTT` in the case that some of the last probes sent out fail. I will keep in mind to think about how to best optimize the implementation here.